src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -1466,7 +1466,16 @@ private TypeDeclarationSyntax ParseClassOrStructOrInterfaceOrRoleDeclaration(Syn
             var name = this.ParseIdentifierToken();
             var typeParameters = this.ParseTypeParameterList();
 
-            var paramList = keyword.Kind == SyntaxKind.RecordKeyword && CurrentToken.Kind == SyntaxKind.OpenParenToken
+            SyntaxToken? forKeyword = null;
+            TypeSyntax? underlyingType = null;
+            if (mainKeyword.Kind == SyntaxKind.ExtensionKeyword
+                && CurrentToken.Kind == SyntaxKind.ForKeyword)
+            {
+                forKeyword = EatToken(SyntaxKind.ForKeyword);
+                underlyingType = ParseType();
```

Comment:
consider calling this extensionUnderlyingType to make it clear this is just for extensions.